tonic/src/transport/server/mod.rs

```diff
@@ -527,42 +536,213 @@ impl<L> Server<L> {
+                            continue;
+                        },
+                        None => {
+                            break
+                        },
+                    };
+
+                    trace!("connection accepted");
+
+                    poll_fn(|cx| svc.poll_ready(cx))
+                        .await
+                        .map_err(super::Error::from_source)?;
+
+                    let req_svc = svc
+                        .call(&io)
+                        .await
+                        .map_err(super::Error::from_source)?;
+                    let hyper_svc = TowerToHyperService::new(req_svc);
+
+                    serve_connection(io, hyper_svc, server.clone(), graceful.then(|| signal_rx.clone()));
+                }
+            }
+        }
+
+        if graceful {
+            let _ = signal_tx.send(());
+            drop(signal_rx);
+            trace!(
+                "waiting for {} connections to close",
+                signal_tx.receiver_count()
+            );
+
+            // Wait for all connections to close
+            signal_tx.closed().await;
         }
 
         Ok(())
     }
 }
 
+// This is moved to its own function as a way to get around
+// https://github.com/rust-lang/rust/issues/102211
+fn serve_connection<IO, S>(
+    io: ServerIo<IO>,
+    hyper_svc: TowerToHyperService<S>,
+    builder: ConnectionBuilder,
+    mut watcher: Option<tokio::sync::watch::Receiver<()>>,
+) where
+    S: Service<Request<BoxBody>, Response = Response<BoxBody>> + Clone + Send + 'static,
+    S::Future: Send + 'static,
+    S::Error: Into<BoxError> + Send,
+    IO: AsyncRead + AsyncWrite + Connected + Unpin + Send + 'static,
+    IO::ConnectInfo: Clone + Send + Sync + 'static,
+{
+    tokio::spawn(async move {
+        {
+            let mut sig = pin!(Fuse {
+                inner: watcher.as_mut().map(|w| w.changed()),
+            });
+
+            let mut conn = pin!(builder.serve_connection(TokioIo::new(io), hyper_svc));
+
+            loop {
+                tokio::select! {
+                    rv = &mut conn => {
+                        if let Err(err) = rv {
+                            debug!("failed serving connection: {:#}", err);
+                        }
+                        break;
+                    },
+                    _ = &mut sig => {
+                        conn.as_mut().graceful_shutdown();
+                    }
+                }
+            }
+        }
+
+        drop(watcher);
+        trace!("connection closed");
+    });
+}
+
+type ConnectionBuilder = hyper_util::server::conn::auto::Builder<TokioExecutor>;
+
+/// An adaptor which converts a [`tower::Service`] to a [`hyper::service::Service`].
+///
+/// The [`hyper::service::Service`] trait is used by hyper to handle incoming requests,
+/// and does not support the `poll_ready` method that is used by tower services.
+#[derive(Debug, Copy, Clone)]
+pub struct TowerToHyperService<S> {
```

Comment:
Do we need to make these public? We could in theory just accept `tower::Service` and convert under the hood?